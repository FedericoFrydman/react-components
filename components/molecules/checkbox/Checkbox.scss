@import '../../assets/styles/colors.scss';

.lsux-cb-container {
  align-items: center;
  display: flex;
  margin: 2px;

  & input {
    height: 0;
    margin: 0;
    opacity: 0;
    width: 0;
  }

  body.using-key input:focus ~ &__cb {
    outline: 2px solid var(--checkbox-cb-enabled-focus-halo);
  }

  &__cb {
    align-items: center;
    border: 2px solid var(--checkbox-cb-enabled-unchecked-border);
    border-radius: 3px;
    cursor: pointer;
    display: flex;
    height: 22px;
    justify-content: center;
    min-width: 22px;
    transition: all 0.1s ease-in-out;

    img {
      display: none;
      height: 18px;
      user-select: none;
      width: 18px;
    }

    &--filled--error,
    &--outlined--error {
      border: 2px solid $R200;
    }
  }

  input:checked ~ &__cb {
    background-color: var(--checkbox-cb-enabled-checked-filled-background);
    border: 2px solid var(--brand-color-prim200);

    &--filled {
      background-color: var(--checkbox-cb-enabled-checked-filled-background);
      border-color: var(--checkbox-cb-enabled-checked-filled-border);
      img {
        display: block;
        filter: $filter-white;
      }
      &--error {
        background-color: $R200;
        border: 2px solid $R200;

        img {
          display: block;
          filter: $filter-white;
        }
      }
    }

    &--outlined {
      background-color: var(--checkbox-cb-enabled-checked-outlined-background);
      img {
        display: block;
        filter: var(--brand-color-prim200-filter);
      }
      &--error {
        background-color: $N00;
        border: 2px solid $R200;
        img {
          display: block;
          filter: $filter-red-200;
        }
      }
    }
  }

  input:disabled ~ &__cb {
    border: 2px solid $N500;
    cursor: not-allowed;
  }
  input:disabled:checked ~ &__cb {
    cursor: not-allowed;

    &--filled,
    &--filled--error {
      background-color: var(--checkbox-cb-disabled-checked-filled-background);
      border-color: var(--checkbox-cb-disabled-checked-filled-border);
      img {
        display: block;
        filter: $filter-white;
      }
    }

    &--outlined,
    &--outlined--error {
      background-color: var(--checkbox-cb-disabled-checked-outlined-background);
      border-color: var(--checkbox-cb-disabled-checked-outlined-border);
      img {
        display: block;
        filter: $filter-gray;
      }
    }

    img {
      display: block;
      filter: $filter-gray;
    }
  }

  input:disabled ~ .lsux-text {
    color: $N500;
    cursor: not-allowed;
  }
}
