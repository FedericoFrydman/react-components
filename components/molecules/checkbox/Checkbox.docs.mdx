import { ArgsTable, Story, Canvas } from '@storybook/addon-docs';
import { Checkbox } from './Checkbox';
import { FormField } from '../../organisms/form-field/FormField';
import { Alert } from '../alert/Alert';

# Checkbox

Checkbox allows multiple selections from a set of options. User can select none, one or multiple.
Checkbox component is always paired with a label. It can be single, as a group, or nested.
Checkbox recieves an onClick prop in order to handle preventDefault or stopPropagation functions. The onClick prop is not required, and it should work without it.

**This documentantion is just to display the input props where you can interact with them and see how it behaves. For implementation please use [FormField](/?path=/docs/forms-form-fields--checkbox-field) docs to gide.**

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  The Input component must be used as a child of a FormField component.
</Alert>

<Canvas>
  <Story id="controls-form-fields-checkbox-props--unchecked" />
</Canvas>

<ArgsTable of={Checkbox} />

## Example Code

```tsx
<FormField id="checox-field">
  <Checkbox rightLabel="Checkbox Label" name="checkbox-field" checked={true} onChange={() => null} readOnly />
</FormField>
```

### Result

<FormField id="checox-field">
  <Checkbox rightLabel="Checkbox Label" name="checkbox-field" checked={true} onChange={() => null} readOnly />
</FormField>

<p style={{ margin: 0 }}>&nbsp;</p>

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  The readOnly property here is just to avoid the ReactJS warning that indicates all checkox should have an onChange to
  handle the value state. Please use an onChange function.
</Alert>

<p style={{ margin: 0 }}>&nbsp;</p>

## Types

<p style={{ margin: 0 }}>&nbsp;</p>

### Checkbox group

Checkboxes can be grouped for users to pick from a set of related options. _(This is just the way it could be disabled)_

<Canvas>
  <Story id="controls-form-fields-checkbox-props--group" />
</Canvas>

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  Please use ReactJS Hook to handle the state for each checkbox.
</Alert>

### Nested Checkboxes

Checkboxes may be nested to show a hierarchy of selections. Selecting a parent will automatically activate and select all children.

Deselecting a parent checkbox will deactivate and deselect all children. If a group of child checkboxes contains a mixture of checked and unchecked, show an indeterminate state with a dash.

<Canvas>
  <Story id="controls-form-fields-checkbox-props--nested" />
</Canvas>

## Variants

The `Checkbox` component now has two variants: `outlined` and `filled`. The `filled` variant is the default.

### Outlined

<Canvas>
  <Story id="controls-form-fields-checkbox-props--checked-outlined" />
</Canvas>

### Filled

<Canvas>
  <Story id="controls-form-fields-checkbox-props--checked-filled" />
</Canvas>

## Guidelines

<p style={{ margin: 0 }}>&nbsp;</p>

- **Do** Always provide a label.
- **Do** Write labels in sentence case.
- **Do not** Use punctuation in labels or after checkbox fields.
- **Do** Keep labels short so users can select quickly and confidently.
- **Do** Write labels as statements that the selected state makes true and the unselected state makes false.
- **Do** Title a checkbox group if the content of a group of checkboxes is not immediately apparent, provide a title to clarify.
- **Do** Both the checkbox and label should trigger selection.

<Alert title="Dev Note" severity="warning" icon="alert_warning">
  Remember, when you make a list of components, you must add a unique key to each component.
</Alert>

## States

### Disbled

<Canvas>
  <Story id="controls-form-fields-checkbox-props--disabled-filled" />
</Canvas>

### Mixed

<Canvas>
  <Story id="controls-form-fields-checkbox-props--mixed-filled" />
</Canvas>

### Error

<Canvas>
  <Story id="controls-form-fields-checkbox-props--error" />
</Canvas>
