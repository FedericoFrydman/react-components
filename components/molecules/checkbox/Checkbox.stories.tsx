import { FC } from 'react';
import { Story, Meta } from '@storybook/react';

import { Checkbox, CheckboxProps } from './Checkbox';

import docs from './Checkbox.docs.mdx';
import '../../assets/styles/utilities.scss';

const CheckboxStory: Story<CheckboxProps> = (args: CheckboxProps) => <Checkbox {...args} />;

export const Unchecked = CheckboxStory.bind({});
Unchecked.args = {
  name: 'unchecked-checkbox',
  onChange: () => alert('I was clicked!'),
  variant: 'filled',
};

export const CheckedFilled = CheckboxStory.bind({});
CheckedFilled.args = {
  checked: true,
  name: 'checkedFilled-checkbox',
  onChange: () => alert('I was clicked!'),
  variant: 'filled',
};

export const CheckedOutlined = CheckboxStory.bind({});
CheckedOutlined.args = {
  checked: true,
  name: 'checkedOutlined-checkbox',
  onChange: () => alert('I was clicked!'),
  variant: 'outlined',
};

export const MixedFilled = CheckboxStory.bind({});
MixedFilled.args = {
  checked: true,
  mixed: true,
  name: 'mixed-checkbox',
  onChange: () => alert('I was clicked!'),
  variant: 'filled',
};

export const MixedOutlined = CheckboxStory.bind({});
MixedOutlined.args = {
  checked: true,
  mixed: true,
  name: 'mixed-checkbox',
  onChange: () => alert('I was clicked!'),
  variant: 'outlined',
};

export const WithLabel = CheckboxStory.bind({});
WithLabel.args = {
  checked: true,
  name: 'mixed-checkbox',
  onChange: () => alert('I was clicked!'),
  rightLabel: 'Example Label',
};

export const WithOnClick = CheckboxStory.bind({});
WithOnClick.args = {
  onChange: () => alert('I was clicked!'),
  onClick: () => alert('I was clicked!'),
  rightLabel: 'I have an onClick fn',
};

export const DisabledFilled = CheckboxStory.bind({});
DisabledFilled.args = {
  checked: true,
  disabled: true,
  name: 'disable-checkbox',
  onChange: () => alert('I was clicked!'),
  rightLabel: 'Label Text',
  variant: 'filled',
};

export const DisabledOutlined = CheckboxStory.bind({});
DisabledOutlined.args = {
  checked: true,
  disabled: true,
  name: 'disable-checkbox',
  onChange: () => alert('I was clicked!'),
  rightLabel: 'Label Text',
  variant: 'outlined',
};

export const Error = CheckboxStory.bind({});
Error.args = {
  checked: true,
  error: true,
  name: 'error-checkbox',
  onChange: () => alert('I was clicked!'),
  rightLabel: 'Label Text',
};

export const Group: FC = () => (
  <div>
    <Checkbox
      onChange={() => alert('I was clicked!')}
      key={0}
      name="checkbox-one"
      checked={true}
      rightLabel="Jacob Jones"
      readOnly
    />
    <Checkbox
      onChange={() => alert('I was clicked!')}
      key={1}
      name="checkbox-two"
      rightLabel="Arlene McCoy"
      classNames={['mt-4']}
      readOnly
    />
    <Checkbox
      onChange={() => alert('I was clicked!')}
      key={2}
      name="checkbox-three"
      rightLabel="Devon Lane"
      classNames={['mt-4']}
      readOnly
    />
    <Checkbox
      onChange={() => alert('I was clicked!')}
      key={3}
      name="checkbox-four"
      rightLabel="Courtney Henry"
      classNames={['mt-4']}
      readOnly
    />
  </div>
);

export const Nested: FC = () => (
  <div style={{ display: 'flex' }}>
    <div className="mr-6">
      <Checkbox
        onChange={() => alert('I was clicked!')}
        name="checkbox-select"
        checked={true}
        rightLabel="Select all"
        readOnly
      />
      <div className="ml-4">
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-opt-one"
          rightLabel="Option 1"
          classNames={['mt-4']}
          checked={true}
          readOnly
        />
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-opt-two"
          rightLabel="Option 2"
          classNames={['mt-4']}
          checked={true}
          readOnly
        />
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-opt-three"
          rightLabel="Option 3"
          classNames={['mt-4']}
          checked={true}
          readOnly
        />
      </div>
    </div>
    <div>
      <Checkbox
        onChange={() => alert('I was clicked!')}
        name="checkbox-deselect"
        mixed={true}
        checked={true}
        rightLabel="Deselect all"
        readOnly
      />
      <div className="ml-4">
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-des-one"
          rightLabel="Option 1"
          classNames={['mt-4']}
          checked={true}
          readOnly
        />
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-des-two"
          rightLabel="Option 2"
          classNames={['mt-4']}
          readOnly
        />
        <Checkbox
          onChange={() => alert('I was clicked!')}
          name="checkbox-des-three"
          rightLabel="Option 3"
          classNames={['mt-4']}
          checked={true}
          readOnly
        />
      </div>
    </div>
  </div>
);

export default {
  argTypes: {
    checked: { control: { type: 'boolean' } },
    disabled: { control: { type: 'boolean' } },
    error: { control: { type: 'boolean' } },
    leftLabel: { control: { type: 'text' } },
    mixed: { control: { type: 'boolean' } },
    name: { control: { type: 'text' } },
    rightLabel: { control: { type: 'text' } },
    variant: {
      control: { type: 'select' },
      options: ['filled', 'outlined'],
    },
  },
  component: CheckboxStory,
  parameters: {
    docs: { page: docs },
  },
  title: 'Controls/Form Fields/Checkbox Props',
} as Meta;
