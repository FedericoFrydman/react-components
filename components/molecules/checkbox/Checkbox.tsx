import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Icon } from '../../../components/atoms/icon/Icon';
import { Text } from '../../../components/atoms/text/Text';
import './Checkbox.scss';

export type CheckboxVariant = 'filled' | 'outlined';

export interface CheckboxProps extends Omit<React.HTMLProps<HTMLInputElement>, 'css' | 'name'> {
  /**
   * Determines if input is checked or not.
   * @default false
   */
  checked?: boolean;
  /**
   * Determines if the input can be of mixed state ('minus' icon when checked).
   * @default false
   */
  mixed?: boolean;
  /**
   * Determines if the checkbox displays an error state.
   * @default false
   */
  error?: boolean;
  /**
   * Determines if the checkbox is disabled or not.
   * @default false
   */
  disabled?: boolean;
  /**
   * Left label text for the checkbox.
   */
  leftLabel?: string;
  /**
   * Right label text for the checkbox.
   */
  rightLabel?: string;
  /**
   * Style of the checkbox. Either filled or outlined.
   * @default filled
   */
  variant?: CheckboxVariant;
  /**
   * Name of the input.
   */
  name: string;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLInputElement>;
}

const CheckboxComponent = (
  {
    checked = false,
    classNames = [],
    disabled = false,
    error = false,
    leftLabel,
    mixed = false,
    name,
    rightLabel,
    onClick,
    variant = 'filled',
    ...props
  }: CheckboxProps,
  ref: CheckboxProps['ref'],
) => {
  const classCheckboxContainer = 'lsux-cb-container';
  const classCheckboxElement = `${classCheckboxContainer}__cb`;
  const cnCheckbox = cn(classCheckboxElement, {
    [`${classCheckboxElement}--mixed`]: mixed,
    [`${classCheckboxElement}--${variant}`]: variant,
    [`${classCheckboxElement}--${variant}${error ? '--error' : ''}`]: variant,
    ['ml-3']: leftLabel,
    ['mr-3']: rightLabel,
  });

  return (
    <label onClick={onClick} className={cn('lsux-cb-container', classNames)} data-testid="lsux-cb-container">
      <input
        {...props}
        onClick={(e) => e.stopPropagation()}
        ref={ref}
        type="checkbox"
        checked={checked}
        name={name}
        disabled={disabled}
        className="lsux-cb-container__input"
      />
      {leftLabel && <Text text={leftLabel} />}
      <div className={cnCheckbox} data-testid="lsux-cb-container__icon">
        {checked && <Icon name={mixed ? 'edit_remove_minus' : 'interface_check'} alt={name} size="large" />}
      </div>
      {rightLabel && <Text text={rightLabel} />}
    </label>
  );
};

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(CheckboxComponent);
