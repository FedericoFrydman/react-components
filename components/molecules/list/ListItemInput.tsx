import { FC, forwardRef, Ref } from 'react';
import cn from 'classnames';

import './list-item-input.scss';
import { Radio, RadioProps } from '../radio/Radio';
import { Checkbox, CheckboxProps } from '../checkbox/Checkbox';
import { ListItem, ListItemProps, ListItemText } from './ListItem';
import { Switch, SwitchProps } from '../switch/Switch';
import { TextProps } from 'components/atoms/text/Text';

type LIInputElemProps = CheckboxProps | RadioProps | SwitchProps;

const classLIInputControl = 'lsux-list-item-input__control';

export const ListItemInputCheckboxComponent = (props: ListItemInputProps<CheckboxProps>) => {
  const { inputProps, ...otherProps } = props;

  function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    e.preventDefault();
    if (e.target.checked) {
      otherProps.handleClick(e.currentTarget?.id);
    } else {
      otherProps.handleClick(null);
    }
    otherProps.handleClick(e.currentTarget?.id);
  }

  return (
    <ListItemInput {...otherProps} classNames={[classLIInputControl, ...(props?.classNames ?? [])]}>
      <Checkbox {...inputProps} checked={props.isSelected} onChange={handleChange} />
    </ListItemInput>
  );
};

export const ListItemInputRadioComponent = ({ ...props }: ListItemInputProps<RadioProps>) => {
  const { inputProps, ...otherProps } = props;

  function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    e.preventDefault();
    otherProps.handleClick(e.currentTarget?.id);
  }

  return (
    <ListItemInput {...otherProps} classNames={[classLIInputControl, ...(props?.classNames ?? [])]}>
      <Radio {...inputProps} checked={props.isSelected} onChange={handleChange} />
    </ListItemInput>
  );
};

export const ListItemInputSwitchComponent = ({ ...props }: ListItemInputProps<SwitchProps>) => {
  const { inputProps, ...otherProps } = props;
  return (
    <ListItemInput {...otherProps} classNames={[classLIInputControl, ...(props?.classNames ?? [])]}>
      <Switch {...inputProps} isChecked={props.isSelected} onClick={(e) => props.handleClick(e.currentTarget?.id)} />
    </ListItemInput>
  );
};

export interface ListItemInputProps<T> extends ListItemProps {
  /**
   * The ID of the ListItemInput.
   */
  id?: string;
  /**
   * The children of the ListItemInput.
   */
  children?: React.ReactNode | React.ReactNode[];
  /**
   * Primary text for the ListItemInput.
   */
  primary?: string;
  /**
   * Props for the primary text
   * @default {}
   */
  primaryProps?: TextProps;
  /**
   * Secondary text for the ListItemInput.
   */
  secondary?: string;
  /**
   * Props for the secondary text
   * @default {}
   */
  secondaryProps?: TextProps;
  /**
   * Determines the position of the control.
   * @default 'left'
   */
  controlPosition?: 'left' | 'right';
  /**
   * Determines if the ListItemInput is selected.
   */
  isSelected?: boolean;
  /**
   * Determines if the halo around the input is displayed when selected.
   */
  showHalo?: boolean;
  /**
   * Callback function for when the ListItemInput is clicked.
   */
  handleClick?: (id: string) => void;
  /**
   * Props for the underlying input (checkbox, radio, switch)
   */
  inputProps?: T;
  /**
   * Additional classNames to add to the list item.
   * @default []
   */
  classNames?: string[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLLIElement>;
}

const ListItemInputComponent = forwardRef<HTMLLIElement, ListItemInputProps<LIInputElemProps>>(
  (
    {
      children,
      controlPosition = 'left',
      id,
      isSelected,
      primary,
      primaryProps,
      secondary,
      secondaryProps,
      showHalo,
      handleClick,
      classNames = [],
      ...props
    }: ListItemInputProps<LIInputElemProps>,
    ref: ListItemInputProps<LIInputElemProps>['ref'],
  ) => {
    const classLIInput = 'lsux-list-item-input';
    const cnListItemInput = cn({ [`${classLIInput}`]: true }, ...classNames);

    const classLIInputButton = `${classLIInput}__button`;
    const cnLIInputButton = cn({
      [`${classLIInputButton}`]: true,
      [`${classLIInputButton}--selected`]: isSelected,
      [`show-halo`]: showHalo,
    });

    const listItemText = (
      <ListItemText
        primary={primary}
        primaryProps={primaryProps}
        secondary={secondary}
        secondaryProps={secondaryProps}
      />
    );

    return (
      <ListItem {...props} ref={ref} classNames={[cnListItemInput]}>
        <a className={cnLIInputButton} onClick={() => handleClick(id)}>
          {controlPosition === 'right' ? listItemText : null}
          <div className={cn(`${classLIInputButton}__children`, `${classLIInputButton}__children--${controlPosition}`)}>
            {children}
          </div>
          {controlPosition === 'left' ? listItemText : null}
        </a>
      </ListItem>
    );
  },
) as FC<ListItemInputProps<LIInputElemProps>>;

ListItemInputComponent.displayName = 'Modal';

type LIInputCheckboxComponent = React.FC<ListItemInputProps<CheckboxProps>>;
type LIInputRadioComponent = React.FC<ListItemInputProps<RadioProps>>;
type LIInputSwitchComponent = React.FC<ListItemInputProps<SwitchProps>>;
type LIInputComp = React.FC<ListItemInputProps<LIInputElemProps>> & {
  Checkbox: LIInputCheckboxComponent;
  Radio: LIInputRadioComponent;
  Switch: LIInputSwitchComponent;
};

export const ListItemInput = ListItemInputComponent as LIInputComp;

ListItemInput.Checkbox = ListItemInputCheckboxComponent;
ListItemInput.Radio = ListItemInputRadioComponent;
ListItemInput.Switch = ListItemInputSwitchComponent;
