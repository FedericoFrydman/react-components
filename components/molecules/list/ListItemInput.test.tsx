import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { ListItemInput, ListItemInputProps } from './ListItemInput';
import { CheckboxProps } from '../checkbox/Checkbox';

describe('ListItemInput Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test list item';
    wrapper = shallow(
      <ListItemInput.Checkbox
        inputProps={{ name: testId }}
        primary={`$10.00/mo`}
        secondary="Lorem ipsum dolor sit amet"
      />,
    );
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the component the provided classNames', () => {
          const listItemInput = renderListItemInput({ classNames });

          expect(listItemInput).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  const renderListItemInput = (props: ListItemInputProps<CheckboxProps> = {}): HTMLElement => {
    render(<ListItemInput.Checkbox {...props} data-testid={testId} />);

    return screen.queryByTestId(testId);
  };
});
