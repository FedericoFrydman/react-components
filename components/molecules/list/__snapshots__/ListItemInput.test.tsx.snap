// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemInput Component should render the component 1`] = `
<Modal
  classNames={
    Array [
      "lsux-list-item-input__control",
    ]
  }
  primary="$10.00/mo"
  secondary="Lorem ipsum dolor sit amet"
>
  <ForwardRef(CheckboxComponent)
    name="test list item"
    onChange={[Function]}
  />
</Modal>
`;
